File: 0349-intersection-of-two-arrays/0349-intersection-of-two-arrays.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */

function binarySearch(nums, target) {
    let left = 0, right = nums.length - 1;
    while (left <= right) {
        let mid = Math.floor(left + (right - left) / 2);
        if (nums[mid] === target) return true;
        else if (nums[mid] < target) left = mid + 1;
        else right = mid - 1;
    }
    return false;
}

var intersection = function(nums1, nums2) {
    nums1.sort((a, b) => a - b);
    const resultSet = new Set();
    for (const target of nums2) {
        if (binarySearch(nums1, target)) {
            resultSet.add(target);
        }
    }
    return Array.from(resultSet);
};

File: 0349-intersection-of-two-arrays/0349-intersection-of-two-arrays.cpp
class Solution {
public:
    bool binary_search(vector<int>& nums,int target) {
        int left = 0,right = nums.size()-1;
        while(left <= right) {
            int mid = left + (right-left) / 2;
            if(nums[mid] == target) return true;
            else if(nums[mid] < target) left = mid+1;
            else right = mid-1;
        }
        return false;
    }
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(),nums1.end());
        unordered_set<int> result;
        for(int &target: nums2) {
            if(binary_search(nums1,target)) {
                result.insert(target);
            }
        }
        return vector<int> (result.begin(),result.end());
    }
};